// tst-reflect-transformer/dist/contexts/SourceFileContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Context_1 = require("./Context");
const log_1 = require("../log");
const mainVisitor_1 = require("../visitors/mainVisitor");
class SourceFileContext {
    constructor(transformerContext, transformationContext, program, checker, sourceFile) {
        this._typesMetadata = [];
        this._ctorsLibrary = [];
        this.generatedTypeIds = new Set();
        this.log = new log_1.Logger(sourceFile.fileName);
        this.transformerContext = transformerContext;
        this.transformationContext = transformationContext;
        this.program = program;
        this.checker = checker;
        this._currentSourceFile = sourceFile;
        this._context = new Context_1.Context(this, mainVisitor_1.mainVisitor);
    }
    get context() {
        return this._context;
    }
    get typesMetadata() {
        return this._typesMetadata;
    }
    get typesCtors() {
        return this._ctorsLibrary;
    }
    get currentSourceFile() {
        return this._currentSourceFile;
    }
    get metaWriter() {
        return this.transformerContext.metaWriter;
    }
}
exports.default = SourceFileContext;
